File: quiz-data.js
#!/usr/bin/env node
// MongoDB에서 어제의 대화 요약을 가져옴 — Claude가 퀴즈 생성에 사용
const path = require('path');
const { MongoClient } = require('mongodb');
require('dotenv').config({ path: path.join(__dirname, '.env') });

async function main() {
  // 어제 날짜 범위
  const now = new Date();
  const localYesterday = new Date(now);
  localYesterday.setDate(localYesterday.getDate() - 1);
  const yStr = `${localYesterday.getFullYear()}-${String(localYesterday.getMonth() + 1).padStart(2, '0')}-${String(localYesterday.getDate()).padStart(2, '0')}`;
  const start = new Date(yStr + 'T00:00:00.000Z');
  const end = new Date(yStr + 'T23:59:59.999Z');

  const client = new MongoClient(process.env.MONGODB_URI, {
    serverSelectionTimeoutMS: 5000
  });

  try {
    await client.connect();
    const db = client.db('conversation-warehouse');
    const sessions = await db.collection('sessions').find({
      session_date: { $gte: start, $lte: end }
    }).toArray();

    if (sessions.length === 0) {
      // 어제 세션이 없으면 최근 3일 확장
      const threeDaysAgo = new Date(now);
      threeDaysAgo.setDate(threeDaysAgo.getDate() - 3);
      // Use local midnight as lower bound
      const tdStr = `${threeDaysAgo.getFullYear()}-${String(threeDaysAgo.getMonth() + 1).padStart(2, '0')}-${String(threeDaysAgo.getDate()).padStart(2, '0')}`;
      const threeDaysAgoStart = new Date(tdStr + 'T00:00:00.000Z');
      const recent = await db.collection('sessions').find({
        session_date: { $gte: threeDaysAgoStart, $lte: end }
      }).sort({ session_date: -1 }).limit(5).toArray();

      if (recent.length === 0) {
        console.log('No recent sessions found for quiz generation.');
        process.exit(0);
      }
      sessions.push(...recent);
    }

    // 대화 요약 추출 (user/assistant text만, 50K 문자 제한)
    const chunks = [];
    let totalChars = 0;
    const MAX_CHARS = 50000;

    for (const session of sessions) {
      if (totalChars >= MAX_CHARS) break;
      chunks.push(`\n--- Session: ${session.project} (${session.session_date?.toISOString()?.split('T')[0]}) ---\n`);

      for (const msg of session.messages || []) {
        if (totalChars >= MAX_CHARS) break;
        if (msg.role !== 'user' && msg.role !== 'assistant') continue;

        let text = '';
        if (typeof msg.content === 'string') {
          text = msg.content;
        } else if (Array.isArray(msg.content)) {
          text = msg.content.filter(b => b.type === 'text').map(b => b.text).join('\n');
        }

        if (text.length > 0) {
          const prefix = msg.role === 'user' ? 'User' : 'Claude';
          const truncated = text.slice(0, 1000);
          chunks.push(`[${prefix}]: ${truncated}`);
          totalChars += truncated.length;
        }
      }
    }

    console.log(`Found ${sessions.length} sessions from yesterday.`);
    console.log(chunks.join('\n'));

  } catch (err) {
    console.error('Failed to fetch quiz data:', err.message);
    process.exit(1);
  } finally {
    await client.close();
  }
}

main();

File: quiz_data.py
#!/usr/bin/env python3
"""Fetch yesterday's conversation messages from MongoDB for quiz generation.

Filtering is done SERVER-SIDE by an aggregation pipeline that matches on each
MESSAGE's timestamp — not on session_started_at. A long-running session that
started weeks ago but was active yesterday is therefore included (the old
session-level pre-filter silently dropped those — ~86% of real activity on a
typical day). Only the in-window user/assistant messages cross the wire; the DB
does the work, not the client.

Ordering note: the pipeline deliberately has NO $sort stages. Sorting unwound
messages (or the regrouped session docs) server-side carries the full message
content through the sort and blows past MongoDB's 32MB in-memory sort limit
(error 292 / QueryExceededMemoryLimitNoDiskUseAllowed), and allowDiskUse is not
permitted on shared Atlas tiers. Each message's timestamp is carried through the
$group instead, and both orderings — chronological within a session, and
sessions by id — are applied client-side below, where the data is already
truncated and small.
"""

import io
import sys

# Ensure stdout is UTF-8 on Windows
if sys.stdout.encoding != 'utf-8':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')

from _shared import get_db, yesterday_kst_bounds_utc

MAX_CHARS = 200000      # total output budget
MAX_MSG_CHARS = 1000    # per-message truncation


def main():
    # Half-open KST day window [yesterday 00:00, today 00:00) as UTC strings,
    # matched against the per-message UTC `timestamp` (ISO-8601, sorts lexically).
    start_utc, end_utc = yesterday_kst_bounds_utc()

    client, db = get_db()
    try:
        pipeline = [
            # 1) sessions with >=1 message in yesterday's window (NO session_started_at gate)
            {'$match': {'messages.timestamp': {'$gte': start_utc, '$lt': end_utc}}},
            # 2) explode to one doc per message
            {'$unwind': '$messages'},
            # 3) keep only the in-window user/assistant messages
            {'$match': {
                'messages.timestamp': {'$gte': start_utc, '$lt': end_utc},
                'messages.role': {'$in': ['user', 'assistant']},
            }},
            # 4) regroup by session for presentation; timestamp is kept so the
            #    chronological ordering can be applied client-side (see docstring)
            {'$group': {
                '_id': '$session_id',
                'project': {'$first': '$project'},
                'device': {'$first': '$device'},
                'messages': {'$push': {
                    'role': '$messages.role',
                    'content': '$messages.content',
                    'timestamp': '$messages.timestamp',
                }},
            }},
        ]

        groups = list(db['sessions'].aggregate(pipeline))
        if not groups:
            print('No messages found for yesterday for quiz generation.')
            sys.exit(0)

        # Sessions by id (was the trailing server-side $sort)
        groups.sort(key=lambda g: str(g.get('_id') or ''))

        all_chunks = []
        total_msg_count = 0

        for g in groups:
            msgs = g.get('messages') or []
            # Chronological within the session (was the server-side $sort on
            # messages.timestamp, dropped to stay under the 32MB sort limit)
            msgs.sort(key=lambda m: str(m.get('timestamp') or ''))

            session_chunks = []
            for msg in msgs:
                content = msg.get('content', '')
                text = ''
                if isinstance(content, str):
                    text = content
                elif isinstance(content, list):
                    text = '\n'.join(
                        b.get('text', '') for b in content
                        if isinstance(b, dict) and b.get('type') == 'text'
                    )

                if text:
                    prefix = 'User' if msg.get('role') == 'user' else 'Claude'
                    session_chunks.append(f'[{prefix}]: {text[:MAX_MSG_CHARS]}')
                    total_msg_count += 1

            if session_chunks:
                header = f'\n--- Session: {g.get("project", "?")} [{g.get("device", "?")}] ---\n'
                all_chunks.append(header)
                all_chunks.extend(session_chunks)

        output = '\n'.join(all_chunks)
        if len(output) > MAX_CHARS:
            output = output[:MAX_CHARS]

        session_count = sum(1 for c in all_chunks if c.startswith('\n--- Session'))
        print(f'Found {total_msg_count} messages from yesterday across {session_count} sessions.')
        print(output)

    except Exception as e:
        print(f'Failed to fetch quiz data: {e}', file=sys.stderr)
        sys.exit(1)
    finally:
        client.close()


if __name__ == '__main__':
    main()
